internal/cache/file/cache_handler.go

```diff
@@ -0,0 +1,181 @@
+
+	// Todo - (raj-prince) - this lookUp should not change the LRU order.
+	fileInfo := chr.fileInfoCache.LookUp(fileInfoKeyName)
+	if fileInfo == nil {
+		fileInfo = data.FileInfo{
+			Key:              fileInfoKey,
+			ObjectGeneration: object.Generation,
+			Offset:           0,
+			FileSize:         object.Size,
+		}
+
+		evictedValues, err := chr.fileInfoCache.Insert(fileInfoKeyName, fileInfo)
+		if err != nil {
+			return fmt.Errorf("while inserting into the cache: %v", err)
+		}
+
+		for _, val := range evictedValues {
+			fileInfo := val.(data.FileInfo)
+			err := chr.performPostEvictionWork(&fileInfo)
+			if err != nil {
+				return fmt.Errorf("while performing post eviction of %s object error: %v", fileInfo.Key.ObjectName, err)
+			}
+		}
+	} else {
+		job := chr.jobManager.GetJob(object, bucket)
+		jobStatus := job.GetStatus()
+		if jobStatus.Name != downloader.FAILED {
+			// Move this entry on top of LRU.
+			_ = chr.fileInfoCache.LookUp(fileInfoKeyName)
+		}
+	}
+
+	return nil
+}
+
+// GetCacheHandle creates an entry in fileInfoCache if it does not already exist.
+// It creates FileDownloadJob if not already exist. Also, creates localFilePath
+// which contains the downloaded content. Finally, it returns a CacheHandle that
+// contains the async DownloadJob and the local file handle.
+// Acquires and releases LOCK(CacheHandler.mu)
+func (chr *CacheHandler) GetCacheHandle(object *gcs.MinObject, bucket gcs.Bucket, initialOffset int64) (*CacheHandle, error) {
+	err := chr.addFileInfoEntryInTheCacheIfNotAlready(object, bucket)
+	if err != nil {
+		return nil, fmt.Errorf("while adding the entry in the cache: %v", err)
+	}
+
+	localFileReadHandle, err := chr.createLocalFileReadHandle(object.Name, bucket.Name())
+	if err != nil {
+		return nil, fmt.Errorf("while create local-file read handle: %v", err)
+	}
+
+	return NewCacheHandle(localFileReadHandle, chr.jobManager.GetJob(object, bucket), chr.fileInfoCache, initialOffset), nil
+}
+
+// InvalidateCache removes the entry from the fileInfoCache, cancel the async running job incase,
+// and delete the locally downloaded cached-file.
+// Acquires and releases LOCK(CacheHandler.mu)
+// TODO (raj-prince) to implement.
+func (chr *CacheHandler) InvalidateCache(object *gcs.MinObject, bucket gcs.Bucket) error {
+	fileInfoKey := data.FileInfoKey{
+		BucketName: bucket.Name(),
+		ObjectName: object.Name,
+	}
+	fileInfoKeyName, err := fileInfoKey.Key()
+	if err != nil {
+		return fmt.Errorf("while creating key: %v", fileInfoKeyName)
+	}
+
+	chr.mu.Lock()
+	defer chr.mu.Unlock()
+
+	erasedVal := chr.fileInfoCache.Erase(fileInfoKeyName)
+	if erasedVal != nil {
+		fileInfo := erasedVal.(data.FileInfo)
+		err := chr.performPostEvictionWork(&fileInfo)
+		if err != nil {
+			return fmt.Errorf("while performing post eviction of %s object error: %v", fileInfo.Key.ObjectName, err)
+		}
+	}
+	return nil
+}
+
+// Destroy destroys the internal state of CacheHandler correctly specifically closing any fileHandles.
```

Comment:
I think we may not be able to close filehandles explicitly but we will evict all entries and do clean up for evicted entries

Reply:
I agree. So, we have two options:
1. We will provide Clear() method in lru cache which will return all the entries, and will perform cleanUpEvictedEntries.
2. We can directly call jobManager.destroy and fileInfoCache.Clear() - but in this case, we will not be able to delete individual entry unless until we delete the path as part of jobManager.Remove(). Let's discuss offline.

Reply:
As discussed offline, will go with the jobManager.Destroy() and will delete everything under cacheLocation. But we will call it out.